### PR TITLE
Import custom certs

### PR DESCRIPTION
**Pull request by jimmykarily**
Part of: https://github.com/kairos-io/kairos/issues/2418

We need to document this

### CODE REVIEW

File: cmd/build-uki.go [hunk 1]
```diff
@@ -93,7 +93,8 @@ func NewBuildUKICmd() *cobra.Command {
 				return fmt.Errorf("keys directory does not exist: %s", keysDir)
 			}
 			// Check if the keys directory contains the required files
-			requiredFiles := []string{"db.der", "db.key", "db.auth", "KEK.der", "KEK.auth", "PK.der", "PK.auth", "tpm2-pcr-private.pem"}
+			//requiredFiles := []string{"db.der", "db.key", "db.auth", "KEK.der", "KEK.auth", "PK.der", "PK.auth", "tpm2-pcr-private.pem"}
+			requiredFiles := []string{"db.auth", "KEK.auth", "PK.auth"}
```
Comment (jimmykarily) [inline cmd/build-uki.go:97]: I'm not 100% sure about this. With `auth` files around, the `der` files don't seem to be needed anywhere.

File: cmd/genkey.go [hunk 2]
```diff
@@ -142,6 +162,15 @@ func generateAuthKeys(guid efiutil.EFIGUID, keyPath, keyType string) error {
 		sigdb.AppendDatabase(oemSigDb)
 	}
 
+	// TODO: PK too?
```
Comment (jimmykarily) [inline cmd/genkey.go:165]: I'm not sure if we can enroll the "custom" (exported) PK instead of ours. I'm thinking that maybe it won't allow us to enroll  the KEK one then which will be signed by a key not maching the one signing the PK. I'd rather leave PK out for now and enroll ours, since that's the scenario I've tested.

File: cmd/genkey.go [hunk 3]
```diff
@@ -157,3 +186,65 @@ func generateAuthKeys(guid efiutil.EFIGUID, keyPath, keyType string) error {
...
+
+// prepareCustomCertsDir takes a cert directory with keys as they are exported
+// from the UEFI firmware and prepares them for use with sbctl.
+// The keys are exported in the "authenticated variables" format.
+// The keys are expected to be in the "der" format in a specific directory structure.
+// The given directory should have the following files:
+// - db
+// - dbx
+// - KEK
+// - PK
```
Comment (jimmykarily) [inline cmd/genkey.go:198]: Actually `dbx` and `PK` are not needed. I'll remove them from here.

File: cmd/genkey.go [hunk 4]
```diff
@@ -157,3 +186,65 @@ func generateAuthKeys(guid efiutil.EFIGUID, keyPath, keyType string) error {
...
+		}
+	}
+
+	// create a temporary directory to store the custom certs
+	tmpDir, err := os.MkdirTemp("", "sbctl-custom-certs-*")
+	if err != nil {
+		return "", fmt.Errorf("creating temporary directory: %w", err)
+	}
+
+	// TODO: PK too?
```
Comment (jimmykarily) [inline cmd/genkey.go:215]: same as above.

File: cmd/genkey.go [hunk 5]
```diff
@@ -157,3 +186,65 @@ func generateAuthKeys(guid efiutil.EFIGUID, keyPath, keyType string) error {
...
+			return "", fmt.Errorf("reading custom cert file %s: %w", keyType, err)
+		}
+		f := bytes.NewReader(b)
+		siglist, err := signature.ReadSignatureDatabase(f)
+		if err != nil {
+			return "", fmt.Errorf("reading signature database: %w", err)
+		}
+
+		l.Infof("Converting custom certs (type: %s)\n", keyType)
+		for _, sig := range siglist {
```
Comment (jimmykarily) [inline cmd/genkey.go:228]: Code inspired by this: https://github.com/jimmykarily/go-uefi/blob/817a745e790a676138958ccc0e3c0e849d8c5f2a/cmd/efianalyze/main.go#L147

File: pkg/action/build-uki.go [hunk 6]
```diff
@@ -632,9 +632,9 @@ func (b *BuildUKIAction) imageFiles(sourceDir string) (map[string][]string, erro
 		"loader/entries": {},
 		"loader/keys":    {},
 		"loader/keys/auto": {
-			filepath.Join(b.keysDirectory, "PK.der"),
-			filepath.Join(b.keysDirectory, "KEK.der"),
-			filepath.Join(b.keysDirectory, "db.der"),
+			// filepath.Join(b.keysDirectory, "PK.der"),
+			// filepath.Join(b.keysDirectory, "KEK.der"),
+			// filepath.Join(b.keysDirectory, "db.der"),
```
Comment (jimmykarily) [inline pkg/action/build-uki.go:637]: Same as comment above. Maybe I should undo this change to limit this PR to the relevant stuff?

File: cmd/build-uki.go [hunk 7]
```diff
@@ -93,7 +93,8 @@ func NewBuildUKICmd() *cobra.Command {
 				return fmt.Errorf("keys directory does not exist: %s", keysDir)
 			}
 			// Check if the keys directory contains the required files
-			requiredFiles := []string{"db.der", "db.key", "db.auth", "KEK.der", "KEK.auth", "PK.der", "PK.auth", "tpm2-pcr-private.pem"}
+			//requiredFiles := []string{"db.der", "db.key", "db.auth", "KEK.der", "KEK.auth", "PK.der", "PK.auth", "tpm2-pcr-private.pem"}
```
Comment (Itxaka) [inline cmd/build-uki.go:96]: hwo itts the pcr-private key not needed anymore? You cannot sign the measurements without it?

File: cmd/genkey.go (hunk 2 above)
Comment (Itxaka) [reply cmd/genkey.go:165]: isnt PK signs KEK which in turn signs DB? in that case we may need the PK to always be the user one so we can sign all this stuff

File: pkg/action/build-uki.go (hunk 6 above)
Comment (Itxaka) [reply pkg/action/build-uki.go:637]: I think, that if you do manual importing on the firmware, it accepts the der files IIRC? at least on qemu you have to select those files

File: cmd/build-uki.go [hunk 8]
```diff
@@ -93,7 +93,8 @@ func NewBuildUKICmd() *cobra.Command {
...
+			requiredFiles := []string{"db.auth", "KEK.auth", "PK.auth"}
 			for _, file := range requiredFiles {
```
Comment (Itxaka) [inline cmd/build-uki.go:98]: should we precheck here if the customcerts dir is empty and fail before continuing?

Comment (jimmykarily) [reply cmd/build-uki.go:98]: We can yes. I'll add a check. I'll just check that the directory exists and it has at least the db and KEK files. If the file formats are not correct or anything like that, I'll let it fails later.

File: cmd/genkey.go (hunk 2 above)
Comment (jimmykarily) [reply cmd/genkey.go:165]: yes, that's what I was thinking as well. I'll skip the PK then form custom certs. We'll use ours.

File: cmd/build-uki.go (hunk 8 above)
Comment (jimmykarily) [reply cmd/build-uki.go:98]: sorry, my mistake, this is `build-uki`, it doesn't take any "custom certs" argument. You pass the "custom certs" when you call `genkey`. The `build-uki` command just accepts what `genkey` generated.

File: cmd/build-uki.go (hunk 7 above)
Comment (jimmykarily) [reply cmd/build-uki.go:96]: you are right, this one is needed indeed. I'll keep the rest of them in the check for now. This is kind of irrelevant to this PR.

We only have to make sure the "der" files and the "auth" files, include the same certificats. Currently only the "auth" files include the custom certs (see my other comment below).

File: cmd/genkey.go [hunk 9]
```diff
@@ -17,7 +21,10 @@ import (
 	"github.com/foxboron/sbctl/fs"
 )
 
-const skipMicrosoftCerts = "skip-microsoft-certs-I-KNOW-WHAT-IM-DOING"
+const (
+	skipMicrosoftCertsFlag = "skip-microsoft-certs-I-KNOW-WHAT-IM-DOING"
```
Comment (mudler) [inline cmd/genkey.go:25]: just a minor note: both flag needs documentation, but as I have https://github.com/kairos-io/kairos-docs/pull/168 flying, feel free to update that PR or comment there what's missing and I'll update it to cover these two

File: pkg/action/build-uki.go (hunk 6 above)
Comment (mudler) [reply pkg/action/build-uki.go:637]: maybe it's best to have all of them if it doesn't hurt. might be handy to have them in the installer iso in any case

Comment (jimmykarily) [reply pkg/action/build-uki.go:637]: I'm not sure appending them one after the other is correct: https://stackoverflow.com/a/21185723
Also, `openssl` only decodes the first one found when multiple are in a file.

Comment (jimmykarily) [reply pkg/action/build-uki.go]: There is a function that parses multiple keys: https://github.com/kairos-io/enki/pull/86/files#diff-4c9460930d2508990a466375e10496dbb61115e44bcc6ffb733b0487d7052ba5R200

so this should be ok to do.

File: cmd/genkey.go (hunk 9 above)
Comment (jimmykarily) [reply cmd/genkey.go:26]: I commented on the linked PR. Since enki has it's own repo, we could also document it in the README. Otoh, having docs in multiple locations can lead to outdated docs here and there.

File: cmd/build-uki.go (hunk 1 above)
Comment (jimmykarily) [reply cmd/build-uki.go]: I did the work to append the custom certs also in the der files but when I manually enroll the certs in my Asus PN64 pointing to the der files, I only get one cert imported. I guess it only takes the first one in the file (?). In any case, auto-enrollment with systemd-boot still works, either because it uses the "auth" files or because it parses the der files differently.

It's probably ok to keep this PR in its current state.
